### ISSUE TITLE
Enhancement: Tunein support?

### ISSUE THREAD

**Issue by jtennant** [jishi/node-sonos-http-api#366]
Would it be possible to add support for Tunein?

**Comment by jishi** [comment]
Yes, it should be possible. Are you thinking about searching for station by name and starting it or something else?

**Comment by jtennant** [comment]
Exactly, search for a station by name, then start and stop it. Wasn't sure if there was any particular blocker for this (there doesn't seem to be much of a Tunein API for instance).

**Comment by jishi** [comment]
I recall testing this a while back and I think they have an open, non-authenticated api which works. Someone implemented this in a fork of my web-controller.

**Comment by jtennant** [comment]
Have you got ay pointers? I've had a bit of a look around at the forks (and the rest of Github) and can't really find anything. What the prerequisites for being able to add support for a music service - does it need an API or can you do what you need through the Sonos interface? Thanks!

**Comment by jishi** [comment]
The requirements are an external api that would give you the same references (uri:s) that Sonos requires. Preferably non-authenticated, but it could work with separate authentication. We have no access to the Sonos credentials for individual services unfortunately.

I don't remember which fork, sorry, I can't seem to locate it heller from the fork tree.

**Comment by Kamol81** [comment]
Hi - for Tunein stream i have http string. It looks like that.
Maybe it will help You to solve it out.

`POST /MediaRenderer/AVTransport/Control HTTP/1.1
CONTENT-LENGTH: 382
CONTENT-TYPE: text/xml; charset="utf-8"
SOAPACTION: "urn:schemas-upnp-org:service:AVTransport:1#SetAVTransportURI"

<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/"><s:Body><u:SetAVTransportURI xmlns:u="urn:schemas-upnp-org:service:AVTransport:1"><InstanceID>0</InstanceID><CurrentURI>x-rincon-mp3radio://stream4.nadaje.com/muzo</CurrentURI><CurrentURIMetaData></CurrentURIMetaData></u:SetAVTransport></s:Body></s:Envelope>0x0D0x0A0x0D0x0A`

**Comment by Kamol81** [comment]
[tunein.txt](https://github.com/jishi/node-sonos-http-api/files/596868/tunein.txt)

**Comment by jishi** [comment]
There is no confusion on how to actually start a tune-in track, it's just a matter of free time to implement it :)

Also, your example looks like a custom radio, not a tune-in radio. They have another uri-scheme IIRC

**Comment by sseiber** [comment]
I would be happy to add this. It's something I added to the github.com/bencevens Sonos project. Like @jishi says there's no confusion on how to start/stop the stations from playing. Just have to address how it's all supported (e.g. search, etc.)

Some thoughts:
1) Do you have to support search for all music services? I could see supporting playback if you can get the id on your own, but fail gracefully for searches. For instance, a google custom search pointed to site:Tunein.com would be a way to retrieve ids (e.g. from the open-graph meta data in the search results). But, unless there is a public (non-partner api to get tunein radio stations) it would be hard to map into a serviceDef file.

2) I also noticed that telling Sonos to tune to a radio station takes a long time before returning. Depending on who the client is I've noticed they can timeout waiting for a response (like Amazon lamda functions, or Google assistant integration webhooks). Maybe a flag to wait for a response or not would be something to support.

**Comment by jishi** [comment]
It's even simpler. They have an open api, http://opml.radiotime.com/

The id's that you get from that is the same that you use for the `x-sonosapi-stream:s20308?sid=254&flags=32` uri. In this example, `s20308`. 

For searching, you use `http://opml.radiotime.com/Search.ashx?query=p5%20sthlm`.

**Comment by jishi** [comment]
Merged @sseiber's PR that introduces support for starting tunein-station based on idea. No search yet.
